cupy/core/core.pyx

```diff
@@ -2472,100 +2473,115 @@ cpdef ndarray concatenate_method(tup, int axis):
         raise ValueError('Cannot concatenate from empty tuple')
 
     if not have_same_types:
-        dtype = numpy.find_common_type([a.dtype for a in tup], [])
-    return concatenate(tup, axis, shape, dtype)
+        dtype = numpy.find_common_type([a.dtype for a in arrays], [])
+    return _concatenate(arrays, axis, tuple(shape), dtype)
 
 
-cpdef ndarray concatenate(tup, axis, shape, dtype):
+cpdef ndarray _concatenate(list arrays, Py_ssize_t axis, tuple shape, dtype):
+    cdef ndarray a, ret
+    cdef int i
+    cdef bint all_same_type, same_shape_and_contiguous
+    cdef Py_ssize_t axis_size
+    cdef Py_ssize_t threshold_size = 2 * 1024 * 1024
+
+    if len(arrays) > 8:
```

Comment:
Why is this condition required? When an array is small, is this procedure inefficient?

Reply:
Yes.  this procedure needs host to device copy that takes a few time.